File: hedera-mirror-rest/model/nft.js
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2022 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

class Nft {
  /**
   * Parses nft table columns into object
   */
  constructor(nft) {
    this.accountId = nft.account_id;
    this.createdTimestamp = nft.created_timestamp;
    this.delegatingSpender = nft.delegating_spender;
    this.deleted = nft.deleted;
    this.metadata = nft.metadata;
    this.modifiedTimestamp = nft.modified_timestamp;
    this.serialNumber = nft.serial_number;
    this.spender = nft.spender;
    this.tokenId = nft.token_id;
  }

  static tableAlias = 'nft';
  static tableName = this.tableAlias;

  static ACCOUNT_ID = 'account_id';
  static CREATED_TIMESTAMP = 'created_timestamp';
  static DELEGATING_SPENDER = 'delegating_spender';
  static DELETED = 'deleted';
  static METADATA = 'metadata';
  static MODIFIED_TIMESTAMP = 'modified_timestamp';
  static SERIAL_NUMBER = 'serial_number';
  static SPENDER = 'spender';
  static TOKEN_ID = 'token_id';

  /**
   * Gets full column name with table alias prepended.
   *
   * @param {string} columnName
   * @private
   */
  static getFullName(columnName) {
    return `${this.tableAlias}.${columnName}`;
  }
}

export default Nft;

File: hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/token/Nft.java
/*
 * Copyright (C) 2019-2023 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.common.domain.token;

import com.fasterxml.jackson.annotation.JsonUnwrapped;
import com.hedera.mirror.common.converter.AccountIdConverter;
import com.hedera.mirror.common.domain.UpsertColumn;
import com.hedera.mirror.common.domain.Upsertable;
import com.hedera.mirror.common.domain.entity.EntityId;
import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder(toBuilder = true)
@Entity
@AllArgsConstructor
@NoArgsConstructor
@Upsertable
public class Nft {

    @JsonUnwrapped
    @EmbeddedId
    private NftId id;

    @UpsertColumn(coalesce = "case when deleted = true then null else coalesce({0}, e_{0}, {1}) end")
    @Convert(converter = AccountIdConverter.class)
    private EntityId accountId;

    @Column(updatable = false)
    private Long createdTimestamp;

    @Convert(converter = AccountIdConverter.class)
    @UpsertColumn(coalesce = "{0}")
    private EntityId delegatingSpender;

    private Boolean deleted;

    @Column(updatable = false)
    private byte[] metadata;

    private Long modifiedTimestamp;

    @Convert(converter = AccountIdConverter.class)
    @UpsertColumn(coalesce = "{0}")
    private EntityId spender;

    public Nft(long serialNumber, EntityId tokenId) {
        id = new NftId(serialNumber, tokenId);
    }
}
